crates/fuzzing/src/generators/stacks.rs

```diff
@@ -0,0 +1,331 @@
+        // reuse an instance in the oracle, rather than re-instantiate).
+        run_body
+            .instruction(&Instruction::I32Const(0))
+            .instruction(&Instruction::GlobalSet(stack_len_global));
+
+        // Initialize the fuel global.
+        run_body
+            .instruction(&Instruction::LocalGet(0))
+            .instruction(&Instruction::GlobalSet(fuel_global));
+
+        push_func_to_stack(&mut run_body, run_func);
+
+        // Make sure to check for out-of-fuel in the `run` function as well, so
+        // that we also capture stack traces with only one frame, not just `run`
+        // followed by the first locally-defined function and then zero or more
+        // extra frames.
+        check_fuel(&mut run_body);
+
+        // Call the first locally defined function.
+        run_body.instruction(&Instruction::Call(num_imported_funcs));
+
+        check_fuel(&mut run_body);
+        pop_func_from_stack(&mut run_body);
+
+        function(&mut code, run_body);
+
+        let mut get_stack_body = wasm_encoder::Function::new(vec![]);
+        get_stack_body
+            .instruction(&Instruction::I32Const(0))
+            .instruction(&Instruction::GlobalGet(stack_len_global));
+        function(&mut code, get_stack_body);
+
+        section(&mut module, code);
+
+        return module.finish();
+
+        // Helper that defines a section in the module and takes ownership of it
+        // so that it is dropped and its memory reclaimed after adding it to the
+        // module.
+        fn section(module: &mut wasm_encoder::Module, section: impl wasm_encoder::Section) {
+            module.section(&section);
+        }
+
+        // Helper that defines a function body in the code section and takes
+        // ownership of it so that it is dropped and its memory reclaimed after
+        // adding it to the module.
+        fn function(code: &mut wasm_encoder::CodeSection, mut func: wasm_encoder::Function) {
+            func.instruction(&Instruction::End);
+            code.function(&func);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::prelude::*;
+    use wasmparser::Validator;
+
+    #[test]
+    fn stacks_generates_valid_wasm_modules() {
+        let mut rng = SmallRng::seed_from_u64(0);
+        let mut buf = vec![0; 2048];
+        for _ in 0..1024 {
+            rng.fill_bytes(&mut buf);
+            let u = Unstructured::new(&buf);
+            if let Ok(stacks) = Stacks::arbitrary_take_rest(u) {
+                let wasm = stacks.wasm();
+                validate(&wasm);
+            }
+        }
+    }
+
+    fn validate(wasm: &[u8]) {
+        let mut validator = Validator::new();
+        let err = match validator.validate_all(wasm) {
+            Ok(_) => return,
+            Err(e) => e,
+        };
+        drop(std::fs::write("test.wasm", wasm));
+        if let Ok(text) = wasmprinter::print_bytes(wasm) {
+            drop(std::fs::write("test.wat", &text));
```

Comment:
We generally write failing test cases to `test.wa{t,sm}` for fuzz tests and smoke tests like this so it is easier to debug. They are only written if the test fails, so won't otherwise spew stuff to the file system during normal operation.